### PR TITLE
January 2021 Update

### PR DESCRIPTION
**Pull request by Pythonic-Rainbow**
**+** Added ro.py in `libs.yml`
**+** Many embed that displayed a member's name/discriminator/id now directly mentions the member.
**\*** Code refactor and clean up.
**+** Redesign `?libs`

# moderation.py
A large amount of code has been extracted to 2 methods.

**\*** Fixed a bug which Moderators with no Library Developer role can't invoke moderation commands
^ Many commands now work.

I didn't test ?reason ?duration as I didn't have the database on my working dir.

### CODE REVIEW

**Comment by Pythonic-Rainbow** [comment]
Please note that for this to work, Member intent has to be enabled on Discord DevPortal as demonstrated in #developer_spot.

**Comment by Pythonic-Rainbow** [comment]
Also, @jmk-developer suggested that we might want to redesign the layout of `?libs` as he thinks its too messy.

My current proposal is a layout like this
```
**Python**
[ro.py](their repo link): By jmkdev (jmkdev#0001) and iranathan(iranathan#5611)

...
**Java**
~~[Javablox](repo link): By Pythonic_Rainbow (Rainbow Breathing#7219)~~
```
where libs with a strike means dead libs.

I am thinking, should we keep those dead libs in `libs.yml`?

In the future, we might just migrate the list in `libs.yml` into the database, which should allow us to update the table by typing commands.

File: cogs/moderation.py [hunk 1]
```diff
@@ -184,37 +145,24 @@ async def duration(self, ctx, id, *, _duration):
...
-
-        def is_member(message):
-            return message.author.id == member.id
-
-        await ctx.message.channel.purge(limit=int(amount) + 1, check=is_member)
-        await log_channel.send(embed=emb)
+    async def clean(self, ctx, amount: int, member: discord.Member = None):
+        _, emb, ch = prepare_action(ctx, '')
+        emb.add_field(name='Message (Clean)', value=ch.mention, inline=False)
+        emb.add_field(name="Moderator", value=ctx.message.author.mention, inline=False)
```
Comment (Pythonic-Rainbow) [inline cogs/moderation.py]: I should probably move line 151 to line 165

File: cogs/moderation.py [hunk 2]
```diff
@@ -21,123 +53,54 @@ async def ping_library_developers(self, ctx, title, *, message):
...
+        _, emb, channel = prepare_action(ctx, '')
+        process_action(ctx, 'Warn', emb, member, message, None)
 
         await ctx.message.add_reaction('👍')
-        await log_channel.send(embed=emb)
+        await self.log_ch.send(embed=emb)
 
     @commands.command()
-    @commands.has_role("Library Developer" or "Moderator")
+    @commands.has_role("Moderator")
```
Comment (sbhadr) [inline cogs/moderation.py]: Library Developers can see infractions. This was done on purpose.

File: cogs/moderation.py (hunk 1 above)
Comment (sbhadr) [reply cogs/moderation.py]: If it makes sense, I don't have an issue with it.

**Comment by sbhadr** [comment]
> Please note that for this to work, Member intent has to be enabled on Discord DevPortal as demonstrated in #developer_spot.

Has been enabled.

> Also, @jmk-developer suggested that we might want to redesign the layout of `?libs` as he thinks its too messy.
> 
> My current proposal is a layout like this
> 
> ```
> **Python**
> [ro.py](their repo link): By jmkdev (jmkdev#0001) and iranathan(iranathan#5611)
> 
> ...
> **Java**
> ~~[Javablox](repo link): By Pythonic_Rainbow (Rainbow Breathing#7219)~~
> ```
> 
> where libs with a strike means dead libs.
> 
> I am thinking, should we keep those dead libs in `libs.yml`?
> 
> In the future, we might just migrate the list in `libs.yml` into the database, which should allow us to update the table by typing commands.

I do think that it needs a redesign but, striking out libraries as opposed to not having them in the list takes up a lot of unneeded space for seemingly no reason. I think the best course of action would be a command like:

``?libs <language>`` where it returns all of the channels and links in the under the ``python_channelname`` abbreviation. Not only would it be more versatile but, we wouldn't have to update the list manually as much as compared to having a mod create the actual channel. A ``yml`` file is much more simplistic / stylistic than a json file would be (what we used previously iirc) and I really do not want to go back to using a json file. I'm more-so looking for automated & less manual stuff so that it's easier to integrate and the reliance on restarting the bot entirely to pull latest changes can be dropped especially with static files that could be dynamic such as the ``yml``. Some commands however, deserve to be static so, in that sole case, it's fine. Lmk what you think. Thanks for the pull request & maintenance of this.

File: cogs/moderation.py (hunk 2 above)
Comment (Pythonic-Rainbow) [reply cogs/moderation.py]: e16c1b4

File: cogs/moderation.py (hunk 1 above)
Comment (Pythonic-Rainbow) [reply cogs/moderation.py]: e16c1b4

**Comment by Pythonic-Rainbow** [comment]
If I want to create a new python file, is there anything need to change on `generate.py`?

The new code heavily relies on `moderation.xxx_embed()` functions. Might want to create a new class for it.

**Comment by Pythonic-Rainbow** [comment]
Also can we create a `ERROR_LOG_CHANNEL` in .env?
Seems to ease development cuz i have to keep changing the channel ID

**Comment by Pythonic-Rainbow** [comment]
Code refactor is completed. Now it's just the matter of whether to update this or not.

> `?libs <language>` where it returns all of the channels and links in the under the python_channelname abbreviation.

Can you give me a demo response in Markdown?

**Comment by sbhadr** [comment]
> If I want to create a new python file, is there anything need to change on `generate.py`?
> 
> The new code heavily relies on `moderation.xxx_embed()` functions. Might want to create a new class for it.

Generates static files. If you have a static file that needs to be generated on restart (i.e re-cached) then, it makes sense to add it here. Otherwise, the file will not update. https://github.com/RbxAPI/Docs-Bot/blob/rewrite/generate.py#L57

> Also can we create a `ERROR_LOG_CHANNEL` in .env?
> Seems to ease development cuz i have to keep changing the channel ID

Probably a good idea. 

> Can you give me a demo response in Markdown?

Sure. 

**Karen:** ?libs python
**Docs:** 
> **Python**
> Library A - [Author](https://github.com/profile), [link](https://github), [channel](https://discord.com)
> Library B - [Author](https://github.com/profile), [link](https://github), [channel](https://discord.com)
> Library C - [Author](https://github.com/profile), [link](https://github), [channel](https://discord.com)

**Karen:** ?libs js
**Docs:** 
> **Javascript**
> Library A - [Author](https://github.com/profile), [link](https://github), [channel](https://discord.com)
> Library B - [Author](https://github.com/profile), [link](https://github), [channel](https://discord.com)
> Library C - [Author](https://github.com/profile), [link](https://github), [channel](https://discord.com)

Something like this that is simple but, gets the job done.

**Comment by Pythonic-Rainbow** [comment]
So we are purging dead libs?

**Comment by jmkd3v** [comment]
I hope so 😞

**Comment by Pythonic-Rainbow** [comment]
image
How does this look like

**Comment by tacheometry** [comment]
I think it should only mention the channel name in the field value and only have the repo link & name as the title of each field

**Comment by sbhadr** [comment]
> So we are purging dead libs?

Correct.

> image
> How does this look like

I think that the linked author name is a bit redundant. I think it should just be renamed to ``source`` or ``repo``. 

> I think it should only mention the channel name in the field value and only have the repo link & name as the title of each field

Provide a visual representation as to what this means.

**Comment by tacheometry** [comment]
> Provide a visual representation as to what this means.

image
Just realized, you can't set links as titles.

File: EmbedFactory.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from discord import Embed
```
Comment (sbhadr) [inline EmbedFactory.py]: Rename this file better.

**Comment by Pythonic-Rainbow** [comment]
> I think that the linked author name is a bit redundant. I think it should just be renamed to `source` or `repo`.

btw iirc we linked a dev's GitHub profile because some might not be in our server

image

File: cogs/tagging.py [hunk 4]
```diff
@@ -21,14 +21,14 @@ async def tag(self, ctx, name=None):
                 return
             emb = discord.Embed()
             emb.set_footer(text=f'\t\t\t\t\t\t\tTimestamp: {result["date"]}',
-                           icon_url="https://cdn.discordapp.com/attachments/336577284322623499/683028692133216300/ac6e275e1f638f4e19af408d8440e1d1.png")
+                           icon_url='https://avatars1.githubusercontent.com/u/42101452?s=200&v=4')
```
Comment (sbhadr) [inline cogs/tagging.py:24]: What is this link?

File: cogs/verification.py [hunk 5]
```diff
@@ -7,6 +7,8 @@
 import random
 from datetime import datetime
```
Comment (sbhadr) [inline cogs/verification.py:9]: This is being scrapped for a completely new system that I've been working on for a while now. So, I'll accept this but, in the grand scheme of things, it's not as relevant.

File: generate.py [hunk 6]
```diff
@@ -1,21 +1,21 @@
 import requests
 import os
```
Comment (sbhadr) [inline generate.py:3]: On my local branch, this is actually recursive so, you no longer have to input manual files to reload. I'll be pushing that up soon before I accept this PR.

**Review by sbhadr — CHANGES_REQUESTED** [review]

File: cogs/tagging.py (hunk 4 above)
Comment (Pythonic-Rainbow) [reply cogs/tagging.py:24]: Same icon, just shorter link. Exists in `libs.yml` for quite a long time.

File: EmbedFactory.py (hunk 3 above)
Comment (Pythonic-Rainbow) [reply EmbedFactory.py]: Umm I think this is ok actually, idk, what are your suggestions?

File: generate.py (hunk 6 above)
Comment (Pythonic-Rainbow) [reply generate.py:3]: Does that mean a solution for #29

**Comment by Pythonic-Rainbow** [comment]
Supports multi-author. This should be the final format. @jmkd3v 
image

**Comment by jmkd3v** [comment]
Okay, looks great! Glad we can both get credit for our work. Excited to see this implemented soon!

**Comment by Pythonic-Rainbow** [comment]
Right. Everything is set besides the above "rename EmbedFactory" thing

File: generate.py (hunk 6 above)
Comment (sbhadr) [reply generate.py:3]: Correct.

File: EmbedFactory.py (hunk 3 above)
Comment (sbhadr) [reply EmbedFactory.py]: embed.py

Comment (Pythonic-Rainbow) [reply EmbedFactory.py]: https://github.com/RbxAPI/Docs-Bot/pull/37/commits/eaeee9ca5392b8e6130232ab8958767b0eb76fd5

**Comment by Pythonic-Rainbow** [comment]
Wat why is there conflicting files

**Comment by jmkd3v** [comment]
Sanjay seems to have modified main.py which you also modified?

**Comment by Pythonic-Rainbow** [comment]
> Sanjay seems to have modified main.py which you also modified?

yep just realised he pushed something

File: cogs/channels.py [hunk 7]
```diff
@@ -19,9 +21,7 @@ def get_news_role(ctx, channel: discord.TextChannel = None):
     async def leaderboard(self, ctx):
```
Comment (sbhadr) [inline cogs/channels.py:23]: Disabling this. Creates way too much competition in a place that is naturally bound by a community. I will find a libdev only command that satisfies this.

**Comment by Pythonic-Rainbow** [comment]
Resolved file conflicts.

**Review by sbhadr — APPROVED** [review]

**Comment by sbhadr** [comment]
Completing some internal things to make this PR easier to put into life. Will be a bit.

**Comment by jmkd3v** [comment]
Any update on when this will be pushed to main?

**Comment by sbhadr** [comment]
Couple days. I'm trying to fix / improve some core stuff before I push this in.

**Review by sbhadr — APPROVED** [review]
Looks good on my end. Thanks for the PR. Merging.

**Pull request merged**